src/sim/detail/InitializeTracks.cu

```diff
@@ -0,0 +1,339 @@
+
+            // Initialize the particle state from the secondary
+            ParticleTrackView particle(
+                params.particle, states.particle, ThreadId(thread_id));
+            particle = {firstborn.def_id, firstborn.energy};
+
+            // Keep the parent's geometry state
+            GeoTrackView geo(params.geo, states.geo, ThreadId(thread_id));
+            geo.copy_state(geo, firstborn.direction);
+
+            // Mark the secondary as processed and the track as active
+            --inits.secondary_counts[thread_id];
+            firstborn                  = Secondary::from_cutoff();
+            inits.vacancies[thread_id] = flag_alive();
+        }
+        // The track is dead and did not produce secondaries: store the index
+        // so it can be used later to initialize a new track
+        else
+        {
+            inits.vacancies[thread_id] = thread_id;
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create track initializers on device from primary particles.
+ */
+__global__ void process_primaries_kernel(const span<const Primary> primaries,
+                                         const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < primaries.size())
+    {
+        TrackInitializer& init
+            = inits.initializers[inits.initializers.size() + thread_id];
+
+        // Construct a track initializer from a primary particle
+        init = primaries[thread_id];
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create track initializers on device from secondary particles.
+ */
+__global__ void process_secondaries_kernel(const StatePointers states,
+                                           const ParamPointers params,
+                                           const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < states.size())
+    {
+        // Construct the state accessors
+        GeoTrackView geo(params.geo, states.geo, ThreadId(thread_id));
+        SimTrackView sim(states.sim, ThreadId(thread_id));
+
+        // Offset in the vector of track initializers
+        size_type offset_id = inits.secondary_counts[thread_id];
+
+        for (const auto& secondary : states.interactions[thread_id].secondaries)
+        {
+            // If the secondary survived cutoffs
+            if (secondary)
+            {
+                TrackInitializer& init
+                    = inits.initializers[inits.initializers.size() + offset_id];
+
+                // Store the thread ID of the secondary's parent
+                inits.parent[offset_id] = thread_id;
+
+                // Calculate the track ID of the secondary
+                unsigned int track_id = 1 + inits.track_count + offset_id++;
```

Comment:
Is the `1 +` here correct? For the first secondary of the first primary, I would expect the `TrackId` to be 1 (with a parent ID of 0), but here it would be 2 I think.

Digging around and trying to understand the `track_count` made me realize something else -- we really need independent track counters for each separate *event* in flight because right now the `TrackId` is effectively a *global*  counter, isn't it? Can @pcanal confirm whether the track ID should be on a per-event basis?

Reply:
You're right, we need a separate track counter for each event. The `1 +` comes from numbering the primary track IDs starting at 1 instead of 0, but there are a couple other places where I'm not assigning the track ID correctly yet for secondaries... let me fix up all the track numbering.